rfcs/2021-06-15-shipping-blade.md

```diff
@@ -0,0 +1,355 @@
+
+# Detailed Design
+This problem has multiple pieces to it and in order to come up with a solution I have tried to break it down into questions that we must ask to ourselves and then conclude every question.
+
+Here's the visual representation for decision making of bundling. 
+> You can look at the actual source [file of the diagram here](https://drive.google.com/file/d/1AM0ryQCXMbfP4z2PUWpt86KUU5iE8rvR/view?usp=sharing) if you can't see things properly in attached image
+
+![Blade Bundling](./images/bundling-blade-guideline.png)
+
+## What do we want to actually ship?
+* components
+* tokens/themes(need to decide on the naming)
+* types
+* utils
+
+## How will the consumers import things from blade?
+There are multiple ways that consumers can import components, themes, types, utils from blade. Let's look into it one by one
+
+### 1. `import` individual components
+```js
+/* multi-package */
+import Button from '@razorpay-blade/Button' 
+import Text from '@razorpay-blade/Text'
+import overrideTheme from '@razorpay-blade/overrideTheme'
+import paymentTheme from '@razorpay-blade/paymentTheme'
+import getColorScheme from '@razorpay-blade/getColorScheme'
+import useColorScheme from '@razorpay-blade/useColorScheme'
+```
+
+#### Pros
+* No need to maintain a top level re-exports. Just write the component and be done with it.
+
+```js
+// packages/blade/index.ts
+export * from src/components/Button
+export * from src/components/Text
+export * from src/tokens/theme
+```
+
+#### Cons
+* We have to write the whole `import` statement for every component we want. Not a good DX. Imagine importing 15 components and writing 15 import statements for each of them.
+
+### 2. `import` everything from root package
+```js
+/* mono-package */
+import { Button, Text, overrideTheme, paymentTheme, getColorScheme, useColorScheme } from '@razorpay/blade'
+```
+
+#### Pros
+* Don't have to write the whole import statement for every component we want. If there's an import statement already we just add it to the existing named import statement
+
+```js
+// consumer/src/App.tsx
+import { Button, Text, overrideTheme, paymentTheme } from '@razorpay/blade'
+```
+
+#### Cons
+* Everytime we add anything we need to ensure that we re-export it from the top level re-exports.
+
+```js
+// packages/blade/index.ts
+export * from src/components/Button
+export * from src/components/Text
+export * from src/tokens/theme
+```
+
+### 3. `import` from each category(components, tokens,types, utils) root package
+```js
+/* mono-package with per category import */
+import { Button, Text } from '@razorpay/blade/components'
+import { overrideTheme, paymentTheme } from '@razorpay/blade/theme'
+import { getColorScheme, useColorScheme } from '@razorpay/blade/utils'
+```
+
+#### Pros
+* Better readability and structured imports. We know what we want and from where will it come.
+* No long running import statements with concerns mixed
```

Comment:
Another pro I like about this approach is my IDE will start suggesting me component name when I start typing.

Reply:
which should be fine I don't see a problem here. Anything which is related to theme is presumed to be in theme apart from that everything goes into utils. Like what will you do with useColorScheme directly? As a consumer you need `useTheme` which is already part of theme

Reply:
> No long running import statements with concerns mixed

Not really. After using a lot of components, this is going to grow regardless.

Reply:
Agreed with @chaitanyadeorukhkar. Education effort will be required.
So, imagine if as a consumer I want to use `ThemeProvider`, now it's a component but related to theme, so would I expect it to be in `theme` package or `components` package.

Reply:
Maybe we can reduce the education gap by having 2 packages, one for exposing all components (Button, ThemeProvider) and another for non-components?

Reply:
Okay folks, common! The education effort you're talking about is really just the usage. That applies to everything and every library 🤷‍♂️ 

Atleast if I have categories I know where to import what from
- If you want to use components then it would come from components
- if you want to just use tokens then it would come from tokens
- if you want to just use utils for doing custom components in consumer apps it'll come from utils.

Don't want the concerns to be mixed.

>Not really. After using a lot of components, this is going to grow regardless.
@dhruvdutt the sentence says no long running imports with **concerns mixed** 🙂. See the example below